io/shared/src/main/scala/fs2/io/process/ProcessSpawn.scala

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2013 Functional Streams for Scala
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package fs2.io.process
+
+import cats.effect.kernel.Resource
+
+sealed trait ProcessSpawn[F[_]] {
```

Comment:
FTR I don't think this is a particularly meaningful constraint. I think it's safe to assume that if there is the `ProcessSpawn` constraint, rockets may very well be launched 😁 

I almost ripped it out. But it does do something important: it allows us to abstract over alternate implementations. For example via [NuProcess](https://github.com/brettwooldridge/NuProcess) on the JVM or io_uring/libuv on Native. So that's actually pretty useful.

Reply:
Is it a good idea to make it typeclass though ? I though retrospectively, `Network` being a typeclass was not so great (because there can be several implementations of it). Is there some consensus that it's okay to have a "canonical" implementation that is provided implicitly then ?

Reply:
This is an excellent point. I'm not sure there is any consensus on this topic yet 😅

The current design choices mirror the rest of the library, but indeed that might not be what we want to do.

Reply:
So with that in mind, and if it was me, I'd change the `ProcessBuilder` to have the following overrides : 

```scala
def spawn[F[_]](processSpawn: ProcessSpawn[F]) : Resource[F, Process[F]] 
def spawn[F[_]: Async] : Resource[F, Process[F]]  = spawn(ProcessSpawn.forAsync[F])
def spawnStream[F[_]: Async] : Stream[F, Process[F]]  = spawnStream(ProcessSpawn.forAsync[F])
def spawnStream[F[_]](processSpawn: ProcessSpawn[F]) : Stream[F, Process[F]] = spawnStream(ProcessSpawn.forAsync[F]) 
```

This would reduce the ceremonies around starting subprocesses (which is important for scripting DX), whilst providing the flexibility to switch implementations (in particular for test purposes). As you mentioned, it's reasonably safe to assume that missiles will be launched, and therefore `Async` is a safe constraint to have on the `spawn` method.

Reply:
If we have several implementations, it should probably not be `sealed`? NuProcess and alike sound cool, but I would rather see them as standalone libs implementing this interface than blessed integrations living inside the fs2 repo (dependencies!). So more like `http4s.Client`, just with a reasonable default implementation for free.

Reply:
> it should probably not be `sealed`

Well, it's an interesting question how to do this :) for example in `fs2.io.uring` I've implemented a bunch of `sealed` stuff and I think it works fine. It discourages casual extension, which theoretically leaves the door open for us to add new methods to the interface.

Of course, if we did add new methods, that would break external implementations. But those external dependencies should never be used in libraries anyway, just in applications.
- https://github.com/armanbilge/fs2-io_uring#versioning-and-compatibility